File: PerlLASProc/LASProc/LASProc.pl
#!/usr/bin/perl

use strict;
use warnings;
use Data::Dumper;
use Fcntl qw(:seek);    

use lib '../GeoProc';
use GeoPoint;
use GeoVector;
use GeoPlane;
use GeoPolygon;
use GeoFace;
use Utility;
use GeoPolygonProc;

# Create GeoPolygonProc object
my $p1 = new GeoPoint( - 27.28046,  37.11775,  - 39.03485);
my $p2 = new GeoPoint( - 44.40014,  38.50727,  - 28.78860);
my $p3 = new GeoPoint( - 49.63065,  20.24757,  - 35.05160);
my $p4 = new GeoPoint( - 32.51096,  18.85805,  - 45.29785);
my $p5 = new GeoPoint( - 23.59142,  10.81737,  - 29.30445);
my $p6 = new GeoPoint( - 18.36091,  29.07707,  - 23.04144);
my $p7 = new GeoPoint( - 35.48060,  30.46659,  - 12.79519);
my $p8 = new GeoPoint( - 40.71110,  12.20689,  - 19.05819);
my $ptsArray = [$p1, $p2, $p3, $p4, $p5, $p6, $p7, $p8];
my $polygon = new GeoPolygon($ptsArray);
my $procObj = new GeoPolygonProc($polygon);

# LAS File Process

my $rb;
my $wb;
my $wt;
my $value;

my $rbFileName = "../LASInput/cube.las";
my $wbFileName = "../LASOutput/cube_inside.las";
my $wtFileName = "../LASOutput/cube_inside.txt";

open($rb, "<", $rbFileName)
  || die "can't open $rbFileName: $!";
binmode($rb);

open($wb, ">", $wbFileName)
  || die "can't open $wbFileName: $!";
binmode($wb);

open($wt, ">", $wtFileName)
  || die "can't open $wtFileName: $!";

# read LAS header info

sysseek($rb, 96, 0);

sysread($rb, $value, 4);
my $offset_to_point_data_value = unpack('I', $value);
sysread($rb, $value, 4);
sysread($rb, $value, 1);
sysread($rb, $value, 2);
my $record_len = unpack('v', $value);
sysread($rb, $value, 4);
my $record_num = unpack('L', $value);

sysseek($rb, 131, 0);

sysread($rb, $value, 8);
my $x_scale = unpack('d', $value);
sysread($rb, $value, 8);
my $y_scale = unpack('d', $value);
sysread($rb, $value, 8);
my $z_scale = unpack('d', $value);
sysread($rb, $value, 8);
my $x_offset = unpack('d', $value);
sysread($rb, $value, 8);
my $y_offset = unpack('d', $value);
sysread($rb, $value, 8);
my $z_offset = unpack('d', $value);

# read LAS header
sysseek($rb, 0, 0);
sysread($rb, $value, $offset_to_point_data_value);

# write LAS header
syswrite($wb, $value);

my $insideCount = 0;

my ($i, $record_loc, $xi, $yi, $zi, $x, $y, $z);

for($i=0; $i<$record_num; $i++)
{                            
    $record_loc = $offset_to_point_data_value + $record_len * $i;
    sysseek($rb, $record_loc, 0);   
    sysread($rb, $value, 4);
	$xi = unpack('l', $value);
	sysread($rb, $value, 4);
	$yi = unpack('l', $value);
	sysread($rb, $value, 4);
	$zi = unpack('l', $value);
	$x = $xi * $x_scale + $x_offset;
	$y = $yi * $y_scale + $y_offset;
	$z = $zi * $z_scale + $z_offset;
	
	if ($x > $procObj->{x0} && $x < $procObj->{x1} &&
		$y > $procObj->{y0} && $x < $procObj->{y1} &&
		$z > $procObj->{z0} && $x < $procObj->{z1} )
    {
		if ($procObj->PointInside3DPolygon($x, $y, $z))
		{
			# read LAS Point record
			sysseek($rb, $record_loc, 0);   
			sysread($rb, $value, $record_len);
			
			# write LAS Point record
			syswrite($wb, $value);
			                
            # write text LAS Point record
            $wt->printf("%15.6f %15.6f %15.6f\n", $x, $y, $z);

            $insideCount++;
		}
	}
    
	
}	

# update record_num in LAS header with actual writting point records number	
sysseek($wb, 107, 0);  
syswrite($wb, pack('I', $insideCount));
	
$rb->close();
$wb->close();
$wt->close();

File: CPPLASProc/LASProc/LASProc.cpp
#include <tchar.h>

#include <iostream>
#include <fstream>

#include "GeoPolygonProc.h"
using namespace GeoProc;

GeoPolygonProc GetProcObj()
{
	// Initialize cube 
	GeoPoint CubeVerticesArray[8] =
	{  
		GeoPoint(-27.28046,         37.11775,        -39.03485),
		GeoPoint(-44.40014,         38.50727,        -28.78860),
		GeoPoint(-49.63065,         20.24757,        -35.05160),
		GeoPoint(-32.51096,         18.85805,        -45.29785),
		GeoPoint(-23.59142,         10.81737,        -29.30445),
		GeoPoint(-18.36091,         29.07707,        -23.04144),
		GeoPoint(-35.48060,         30.46659,        -12.79519),
		GeoPoint(-40.71110,         12.20689,        -19.05819)
	};

	// Create polygon object		
	std::vector<GeoPoint> verticesVector( CubeVerticesArray, 
		                  CubeVerticesArray + sizeof(CubeVerticesArray) / sizeof(GeoPoint) );
	GeoPolygon polygonObj = GeoPolygon(verticesVector);

	// Create main process object 
	GeoPolygonProc procObj = GeoPolygonProc(polygonObj);

	return procObj;
}

void ProcLAS(GeoPolygonProc procObj)
{	
	std::ifstream rbFile;
    rbFile.open("..\\LASInput\\cube.las",std::ios::binary|std::ios::in);
	
	std::fstream wbFile;
    wbFile.open("..\\LASOutput\\cube_inside.las",std::ios::binary|std::ios::out);
	
	std::ofstream wtFile;
    wtFile.open("..\\LASOutput\\cube_inside.txt",std::ios::out);
	wtFile.precision(4);
	
	// LAS public header variables
	unsigned long offset_to_point_data_value;
	unsigned long variable_len_num;
	unsigned char record_type_c;
	unsigned short record_len;
	unsigned long record_num;
	double x_scale, y_scale, z_scale;
	double x_offset, y_offset, z_offset;	
	
	// Offset bytes and data types are based on LAS 1.2 Format 
    
	// Read LAS public header
	rbFile.seekg(96);	     
    rbFile.read ((char *)&offset_to_point_data_value, 4);
	rbFile.read ((char *)&variable_len_num, 4);
	rbFile.read ((char *)&record_type_c, 1);
	rbFile.read ((char *)&record_len, 2);
	rbFile.read ((char *)&record_num, 4);
		
	rbFile.seekg(131);
	rbFile.read ((char *)&x_scale, 8);
	rbFile.read ((char *)&y_scale, 8);
	rbFile.read ((char *)&z_scale, 8);
	rbFile.read ((char *)&x_offset, 8);
	rbFile.read ((char *)&y_offset, 8);
	rbFile.read ((char *)&z_offset, 8);
	
	// LAS header buffer
	char *bufHeader = (char *)malloc(offset_to_point_data_value);
		
	// Get LAS header
	rbFile.seekg(0);
	rbFile.read(bufHeader, offset_to_point_data_value);
		
	// Write LAS header 
	wbFile.write(bufHeader, offset_to_point_data_value);
	
	// LAS point coordinates
	double x, y, z;        // LAS point actual coordinates			
	long xi, yi, zi;   // LAS point record coordinates

	// Number of inside points
	int insideCount = 0; 

	// Point record buffer
	char *bufRecord = (char *)malloc(record_len);					

	// Process point records
	for(unsigned int i=0;i<record_num;i++)
	{	

		int record_loc = offset_to_point_data_value + record_len * i;

		rbFile.seekg(record_loc);

		// Record coordinates
		rbFile.read ((char *)&xi, 4);
		rbFile.read ((char *)&yi, 4);
		rbFile.read ((char *)&zi, 4);		

		// Actual coordinates
		x = (xi * x_scale) + x_offset;
		y = (yi * y_scale) + y_offset;
		z = (zi * z_scale) + z_offset;	
															
		// Filter out the points outside of boundary to reduce computing
		if( x>procObj.GetX0() && x<procObj.GetX1() && 
			y>procObj.GetY0() && y<procObj.GetY1() && 
			z>procObj.GetZ0() && z<procObj.GetZ1())
		{					
			// Main Process to check if the point is inside of the cube				
			if(procObj.PointInside3DPolygon(x, y, z))
			{										
				// Write the actual coordinates of inside point to text file
				wtFile << std::fixed << x << " " << std::fixed << y	<< " " << std::fixed << z << std::endl;			

				// Get point record
				rbFile.seekg(record_loc);	
				rbFile.read(bufRecord, record_len);

				// Write point record to binary LAS file
				wbFile.write(bufRecord,  record_len);		
				insideCount++;
			}													
		}
					
	}
							
	// Update total record numbers in output binary LAS file
	wbFile.seekp(107);		
	wbFile.write((char *)&insideCount, 4);
								
	rbFile.close();		
	wbFile.close();
	wtFile.close();	

	free(bufHeader);
	free(bufRecord);		
	
}

int _tmain(int argc, _TCHAR* argv[])
{
	// Create procInst	
	GeoPolygonProc procObj = GetProcObj();
	
	// Process LAS
	ProcLAS(procObj);
	  
	return 0;
}







